File: backend/src/models/RealtimeData.js
import mongoose from "mongoose";

const realtimeDataSchema = new mongoose.Schema({
    userId: { type: String, required: true, index: true },
    heart_rate: { type: Number, required: true },
    spo2: { type: Number },
    stress_level: { type: Number },
    blood_pressure: {
        systolic: { type: Number },
        diastolic: { type: Number },
    },
    steps: { type: Number },
    calories_burned: { type: Number },
    timestamp: { type: Date, required: true },
}, {
    timestamps: true,
    collection: "realtime_data"   
});

realtimeDataSchema.index(
    { userId: 1, timestamp: -1 },
    { unique: true, partialFilterExpression: { userId: { $exists: true }, timestamp: { $exists: true } } }
);

// Default export
const RealtimeData = mongoose.model("RealtimeData", realtimeDataSchema);

export default RealtimeData;

File: agent/models/realtime_data.py
from pydantic import BaseModel, field_validator
from datetime import datetime, timezone

class BloodPressure(BaseModel):
    systolic: int
    diastolic: int


class realtime_data(BaseModel):
    userId: str
    heart_rate: int
    spo2: int
    stress_level: int
    blood_pressure: BloodPressure
    steps: int
    calories_burned: int
    timestamp: datetime

    @field_validator("timestamp", mode="before")
    def parse_ts(cls, v):
        if isinstance(v, datetime):
            return v
        return datetime.fromtimestamp(v / 1000, tz=timezone.utc)
